api/app/Models/PoolCandidate.php

```diff
@@ -834,4 +839,112 @@ public function setApplicationSnapshot()
         $this->profile_snapshot = $profile;
         $this->save();
     }
+
+    public function computeDecisionsPerAssessmentStep()
+    {
+        $decisions = [];
+        $this->load([
+            'pool.assessmentSteps',
+            'assessmentResults',
+        ]);
+
+        foreach ($this->pool->assessmentSteps as $step) {
+            $stepId = $step->id;
+            $hasFailure = false;
+            $hasOnHold = false;
+            $hasToAssess = false;
+
+            $stepResults = $this->assessmentResults()->where('assessment_step_id', $stepId);
```

Comment:
Minor suggestion - I think this change will mean it only runs one database query, and filters the cached results. Since the list shouldn't be that long, I expect this to be faster than multiple DB queries.
```suggestion
            $stepResults = $this->assessmentResults->where('assessment_step_id', $stepId);
```